File: Olympiad/COI/official/2008/glasnici.cpp
#include <cstdio>

int N;
double K;
double D[100000];

int ok( double T ) {
  double x = D[0] + T;
  for (int i = 1; i < N; i++) {
     if( D[i]-T > x+K ) return 0;     
     x = (x + K) <? (D[i] + T);
  }
  return 1;
}

int main( void ) {
   scanf( "%lf", &K );
   scanf( "%d", &N );
   for( int i = 0; i < N; ++i ) scanf( "%lf", &D[i] );
      
   double lo = 0.0, hi = 1e9;
   for( int iter = 0; iter < 100; ++iter ) {
      double mid = (lo+hi)/2.0;
      
      if( ok( mid ) ) hi = mid; else lo = mid;
   }
   printf( "%.3lf\n", lo );
   
   return 0;
}

File: Olympiad/COI/official/2008/glasnici.pas
program glasnici;

var
   N : longint;
   K : real;
   D : array[1..100000] of real;
   lo, hi, mid : real;
   i, iter : longint;

function ok( T : real ) : boolean;
var
   x : real;
   i : longint;
begin
   x := D[1] + T;
   ok := true;
   for i := 2 to N do begin
     if D[i]-T > x+K then ok := false;

     if x + K < D[i] + T then x := x + K else x := D[i] + T;
   end;
end;

begin
   readln( K );
   readln( N );
   for i := 1 to N do readln( D[i] );

   lo := 0;
   hi := 1000000000;
   for iter := 1 to 100 do begin
      mid := (lo+hi) / 2;

      if ok( mid ) then hi := mid else lo := mid;
   end;
   writeln( lo:3:3 );
end.
